plugins/govppmux/govpp_channel.go

```diff
@@ -39,22 +37,25 @@ type retryConfig struct {
 
 // ReceiveReply handles request and returns error if occurred. Also does retry if this option is available.
 func (r *govppRequestCtx) ReceiveReply(reply govppapi.Message) error {
+	var timeout time.Duration
+	maxAttempts := r.retry.attempts
+	if r.retry.timeout > 0 { // Default value is 500ms
+		timeout = r.retry.timeout
+	}
+
 	var err error
 	// Receive reply from original send
-	if err = r.requestCtx.ReceiveReply(reply); err != nil && err == core.ErrNotConnected {
-		if r.retry != nil && r.retry.attempts > 0 {
-			// Set default timeout between retries if not set
-			if r.retry.timeout == 0 {
-				r.retry.timeout = defaultRetryRequestTimeout
-			}
-			// Try to re-sent requests
-			for i := 1; i <= r.retry.attempts; i++ {
-				logrus.DefaultLogger().Warnf("Retrying message %v: %d", r.requestMsg.GetMessageName(), i)
-				ctx := r.sendRequest(r.requestMsg)
-				if err = ctx.ReceiveReply(reply); err == nil {
-					return nil
-				}
-				time.Sleep(r.retry.timeout)
+	if err = r.requestCtx.ReceiveReply(reply); err != nil && err == core.ErrNotConnected && maxAttempts > 0 {
+		// Try to re-sent requests
+		for attemptIdx := 1; attemptIdx <= maxAttempts; attemptIdx++ {
+			logrus.DefaultLogger().Warnf("Retrying message %v, attempt: %d", r.requestMsg.GetMessageName(), attemptIdx)
+			if err = r.sendRequest(r.requestMsg).ReceiveReply(reply); err != nil && err == core.ErrNotConnected && attemptIdx != maxAttempts {
+				time.Sleep(timeout)
```

Comment:
As I'm thinking about it.. shouldn't we sleep before the actual retry? Because now it will try to immediately send the retry on `ErrNotConnected` error, which will immediately send same error again, which practically wastes the first retry, not giving much time for possible reconnect.

Reply:
I move the sleep before the another attempt to connect, it makes things a bit easier